server/src/main/java/org/opensearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -125,9 +131,13 @@ static AutoDateHistogramAggregator build(
      * {@link MergingBucketsDeferringCollector#mergeBuckets(long[])}.
      */
     private MergingBucketsDeferringCollector deferringCollector;
+    private Weight[] filters = null;
+    private DateFieldMapper.DateFieldType fieldType;
 
     protected final RoundingInfo[] roundingInfos;
     protected final int targetBuckets;
+    protected int roundingIdx;
+    protected Rounding.Prepared preparedRounding;
```

Comment:
> Most of the new fields can be final since they're set in the constructor, right?

I thought the same initially, but most of them are set conditionally, and being final they cannot take default value. Hence, can't be made final

> Shouldn't roundingIdx be a local variable in getMinimumRounding?

Again did that way earlier, but the FromSingle implementation needs to know the correct rounding index, else it adds unnecessary buckets in the final output.

Reply:
> I thought the same initially, but most of them are set conditionally, and being final they cannot take default value. Hence, can't be made final

What about doing it like this: https://github.com/msfroh/OpenSearch/commit/0b90666c28998465d9baf20a2d0a1bb7742a0cb7

Then we can also reuse the logic between `DateHistogramAggregator` and `AutoDataHistogramAggregator`.

Reply:
> What about doing it like this: https://github.com/msfroh/OpenSearch/commit/0b90666c28998465d9baf20a2d0a1bb7742a0cb7

While I wouldn't have done so much refactoring just for making it final, I really like how it abstracted out the commonalities. Thanks for sharing this!

I will use this for making the tryFastFilterAggregation common across `DateHistogramAggregator` and `AutoDataHistogramAggregator`

Reply:
Okay, doing the refactoring caused an interesting regression which took me sometime to figure out. Otherwise it worked pretty well. The `minimumRounding` method is responsible for setting the correct `preparedRounding` object, and lambda causes lazy invocation of `minimumRounding` whereas `preparedRounding` is picked eagerly. Hence, I modified to pick `preparedRounding` lazily as well.

```
buildFastFilterContext(
            parent(),
            subAggregators,
            context,
            b -> getMinimumRounding(b[0], b[1]),
            preparedRounding,
```